Listen for interrupt signal and close database cleanly

Synapse should close the database cleanly when CTRL-C is run.

Well, I'll try to run it and ctrl-c a few times, but I don't think it's working.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xb6d317]

goroutine 149 [running]:
github.com/dgraph-io/badger/skl.(*Skiplist).IncrRef(...)
	/home/meyer9/go/src/github.com/dgraph-io/badger/skl/skl.go:86
github.com/dgraph-io/badger.(*DB).getMemTables(0xc0000f4c00, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/go/src/github.com/dgraph-io/badger/db.go:459 +0xd7
github.com/dgraph-io/badger.(*DB).get(0xc0000f4c00, 0xc0099153e0, 0x2d, 0x2d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/meyer9/go/src/github.com/dgraph-io/badger/db.go:491 +0x66
github.com/dgraph-io/badger.(*Txn).Get(0xc00991a580, 0xc0099153b0, 0x25, 0x30, 0x25, 0xc0099153b0, 0x5)
	/home/meyer9/go/src/github.com/dgraph-io/badger/txn.go:459 +0xe7
github.com/phoreproject/synapse/beacon/db.(*BadgerDB).GetBlockForHash(0xc00019c860, 0x2a0abcd72d14a037, 0x6233628c5f638bb7, 0xdde4a32f9bc5fd74, 0x9dedafce23b12fec, 0x0, 0x0, 0x0)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/db/badger.go:71 +0x112
github.com/phoreproject/synapse/beacon.(*Blockchain).GetBlockByHash(...)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/blockchain.go:352
github.com/phoreproject/synapse/beacon.(*ChainView).GetStateBySlot(0xc009a68420, 0x135, 0x1, 0x1, 0xc00ae6a840)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/blockchain.go:566 +0x110
github.com/phoreproject/synapse/primitives.(*State).ProcessBlock(0xc011b46780, 0xc011826d80, 0x16bdd40, 0x1068400, 0xc009a68420, 0x0, 0x0, 0xc009a72000)
	/home/meyer9/go/src/github.com/phoreproject/synapse/primitives/state.go:1969 +0xfbc
github.com/phoreproject/synapse/beacon.(*StateManager).AddBlockToStateMap(0xc0094f2000, 0xc011826d80, 0x6233628c5f638b00, 0xdde4a32f9bc5fd74, 0x9dedafce23b12fec, 0xc011b46300)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/statemanager.go:194 +0x232
github.com/phoreproject/synapse/beacon.(*Blockchain).AddBlockToStateMap(...)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/state.go:28
github.com/phoreproject/synapse/beacon.(*Blockchain).ProcessBlock(0xc00939a9f0, 0xc011826d80, 0x6233628c5f630001, 0xdde4a32f9bc5fd74, 0x9dedafce23b12fec)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/state.go:75 +0x447
github.com/phoreproject/synapse/beacon.SyncManager.handleReceivedBlock(0xc0000b8140, 0x12a05f200, 0x0, 0xc00939a9f0, 0xc011826d80, 0xc009430980, 0xc00a7f9400, 0x0, 0x0)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/syncmanager.go:313 +0x413
github.com/phoreproject/synapse/beacon.SyncManager.onMessageBlock(0xc0000b8140, 0x12a05f200, 0xc000187300, 0xc00939a9f0, 0xc009430980, 0x1063880, 0xc0114a0a00, 0xc0000c8160, 0xd99260)
	/home/meyer9/go/src/github.com/phoreproject/synapse/beacon/syncmanager.go:270 +0x1401
github.com/phoreproject/synapse/p2p.(*HostNode).handleMessage(0xc0000b8140, 0xc009430980, 0x1063880, 0xc0114a0a00, 0x34001, 0xc0094db000)
	/home/meyer9/go/src/github.com/phoreproject/synapse/p2p/hostnode.go:151 +0x3ea
github.com/phoreproject/synapse/p2p.(*HostNode).setupPeerNode.func5.1(0x1063880, 0xc0114a0a00, 0x1063880, 0xc0114a0a00)
	/home/meyer9/go/src/github.com/phoreproject/synapse/p2p/hostnode.go:281 +0x47
github.com/phoreproject/synapse/p2p.processMessages(0xc0000330e0, 0xc00ae6df88, 0xc000050f70, 0xc000050f00)
	/home/meyer9/go/src/github.com/phoreproject/synapse/p2p/messagehandler.go:126 +0x24a
github.com/phoreproject/synapse/p2p.(*HostNode).setupPeerNode.func5(0xc0000818d0, 0xc0000b8140, 0xc009430980, 0xc000033320, 0x4, 0x4)
	/home/meyer9/go/src/github.com/phoreproject/synapse/p2p/hostnode.go:280 +0x96
created by github.com/phoreproject/synapse/p2p.(*HostNode).setupPeerNode
	/home/meyer9/go/src/github.com/phoreproject/synapse/p2p/hostnode.go:279 +0x758
exit status 2
^C
```

Could be a problem with Goroutines not being cleaned up properly. Maybe we should use the context API?

--- a/beacon/blockchain.go
+++ b/beacon/blockchain.go
@@ -563,14 +563,6 @@ func (c *ChainView) GetStateBySlot(slot uint64) (*primitives.State, error) {
 	if err != nil {
 		return nil, err
 	}
-	lastBlock, err := c.blockchain.GetBlockByHash(h)
-	if err != nil {
-		return nil, err
-	}
-	h, err = ssz.TreeHash(lastBlock)
-	if err != nil {
-		return nil, err
-	}
 	state, found := c.blockchain.stateManager.GetStateForHash(h)
 	if !found {
 		return nil, errors.New("could not find state at slot")

--- a/beacon/db/badger.go
+++ b/beacon/db/badger.go
@@ -35,12 +35,19 @@ func NewBadgerDB(databaseDir string) *BadgerDB {
 	opts := badger.LSMOnlyOptions
 	opts.Dir = databaseDir
 	opts.ValueDir = databaseDir
+
+	opts.ValueLogFileSize = 8 << 20
+
 	if runtime.GOOS == "windows" {
 		opts.Truncate = true
-		opts.ValueLogFileSize = 1024 * 1024
+		opts.ValueLogFileSize = 1 << 20
 		// Not sure how this option takes effect.
 		//opts.CompactL0OnClose = false
 	}
+	opts.NumMemtables = 4
+	opts.MaxTableSize = 32 << 20 // 32 MB
+	opts.NumCompactors = 1
+
 	db, err := badger.Open(opts)
 	if err != nil {
 		log.Fatal(err)

